sand/controllers/transform.py

```diff
@@ -0,0 +1,190 @@
+import json
+import re
+import sys
+import traceback
+from typing import Dict, List, Union
+
+from flask import jsonify, request
+from flask.blueprints import Blueprint
+from RestrictedPython import compile_restricted_function, safe_globals
+from werkzeug.exceptions import BadRequest
+
+from sand.models.table import Link, Table, TableRow
+from sand.models.transform import Context
+
+transform_bp = Blueprint("transform", "transform")
+USER_DEFINE_FUNCTION_NAME = ""
+
+
+def transform_map(transform_func, data, tolerance):
+    transformed_data = []
+    exec_tolerance = tolerance
+    for path, value, context in data:
+        value = value[0]
+        tdata = dict()
+        tdata["path"] = path
+        tdata["value"] = value
+        try:
+            result = transform_func(value, context)
+            tdata["ok"] = result
+            transformed_data.append(tdata)
+        except Exception as err:
+            (exc, value, tb) = sys.exc_info()
+            tb = tb.tb_next  # Skip *this* frame
+            tdata["error"] = "".join(traceback.format_exception(exc, value, tb))
+            transformed_data.append(tdata)
+            exec_tolerance -= 1
+            if exec_tolerance == 0:
+                return transformed_data
+    return transformed_data
+
+
+def transform_filter(transform_func, data, tolerance):
+    transformed_data = []
+    exec_tolerance = tolerance
+    for path, value, context in data:
+        tdata = dict()
+        value = value[0]
+        tdata["path"] = path
+        tdata["value"] = value
+        try:
+            result = transform_func(value, context)
+            if type(result) != bool:
```

Comment:
don't do type check like this, use instance of

Reply:
Updated it, thanks